aioaws/sqs.py

```diff
@@ -0,0 +1,177 @@
+from contextlib import AsyncExitStack, asynccontextmanager
+from dataclasses import dataclass
+from typing import Any, AsyncIterator, Iterable, Mapping, Optional, Union
+
+from httpx import AsyncClient, Timeout
+from pydantic import BaseModel, Field
+
+from .core import AWSV4AuthFlow
+
+
+class SQSClientConfig(BaseModel):
+    aws_access_key_id: str
+    aws_secret_key: str
+    aws_region: str
+
+
+class SQSMessage(BaseModel):
+    message_id: str
+    receipt_handle: str
+    md5_of_body: str
+    body: str
+    attributes: Mapping[str, Any]
+
+
+class PollConfig(BaseModel):
+    wait_time: int = Field(10, gt=0)
+    max_messages: int = Field(1, ge=1, le=10)
+
+
+@dataclass
+class _QueueName:
+    name: str
+
+
+@dataclass
+class _QueueURL:
+    url: str
+
+
+MAX_VISIBILITY_TIMEOUT = 12 * 60 * 60  # 12 hours in seconds
+
+
+class SQSClient:
+    def __init__(
+        self,
+        queue_name_or_url: str,
+        config: SQSClientConfig,
+        *,
+        client: AsyncClient,
+    ) -> None:
+        self._queue_name_or_url: Union[_QueueName, _QueueURL]
+        if queue_name_or_url[:4] == 'http':
+            self._queue_name_or_url = _QueueURL(queue_name_or_url)
+        else:
+            self._queue_name_or_url = _QueueName(queue_name_or_url)
+        self._client = client
+        self._auth = AWSV4AuthFlow(
+            aws_access_key_id=config.aws_access_key_id,
+            aws_secret_key=config.aws_secret_key,
+            region=config.aws_region,
+            service='sqs',
+        )
+        self._service_url = f'htpts://sqs.{config.aws_region}.amazonaws.com'
+
+    async def _get_queue_url_from_name_and_region(
+        self,
+        queue_name: str,
+        client: AsyncClient,
+        auth: AWSV4AuthFlow,
+    ) -> str:
+        resp = await client.get(
+            url=self._service_url,
+            params={
+                'GetQueueUrl': 'ListQueues',
+                'QueueName': queue_name,
+            },
+            auth=auth,
+        )
+        resp.raise_for_status()
+        return resp.json()['QueueUrl']
+
+    async def _get_queue_url(self) -> str:
+        if isinstance(self._queue_name_or_url, _QueueName):
+            self._queue_name_or_url = _QueueURL(
+                await self._get_queue_url_from_name_and_region(
+                    self._queue_name_or_url.name,
+                    self._client,
+                    auth=self._auth,
+                )
+            )
+        return self._queue_name_or_url.url
+
+    async def poll(
+        self,
+        *,
+        config: PollConfig = PollConfig(),
```

Comment:
mutable default, better to do `Optional[..] = None ... config = PollConfig()`?

Reply:
addressed